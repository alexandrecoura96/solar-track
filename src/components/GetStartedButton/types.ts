import {TouchableOpacityProps} from 'react-native';

export interface GetStartedButtonProps extends TouchableOpacityProps {
  title: string;
}
