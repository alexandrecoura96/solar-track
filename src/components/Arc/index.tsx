import numeral from 'numeral';
import React from 'react';
import {Dimensions} from 'react-native';
import {RFValue} from 'react-native-responsive-fontsize';
import Svg, {Defs, LinearGradient, Path, Stop, Text} from 'react-native-svg';
import {ArcProps} from './types';

const {width} = Dimensions.get('window');
const size = width - 72;
const strokeWidth = 16;
const strokeWidth2 = 32;
const {PI, cos, sin} = Math;
const r = (size - strokeWidth) / 2;
const cx = size / 2;
const cy = size / 2;
const A = PI + PI * 0.4;
const startAngle = PI + PI * 0.2;
const endAngle = 2 * PI - PI * 0.2;
const x1 = cx - r * cos(startAngle);
const y1 = -r * sin(startAngle) + cy;
const x2 = cx - r * cos(endAngle);
const y2 = -r * sin(endAngle) + cy;
const d = `M ${x1} ${y1} A ${r} ${r} 0 1 0 ${x2} ${y2}`;

const startAngle2 = PI + PI * 0.2;
const endAngle2 = 2 * PI - PI * 0.2;
const x1_2 = cx - (r - strokeWidth2 / 1) * cos(startAngle2);
const y1_2 = -(r - strokeWidth2 / 1) * sin(startAngle2) + cy;
const x2_2 = cx - (r - strokeWidth2 / 1) * cos(endAngle2);
const y2_2 = -(r - strokeWidth2 / 1) * sin(endAngle2) + cy;
const d2 = `M ${x1_2} ${y1_2} A ${r - strokeWidth2 / 1} ${
  r - strokeWidth2 / 1
} 0 1 0 ${x2_2} ${y2_2}`;

export const Arc = ({content}: ArcProps) => {
  const circumference = r * A;
  const α = 100 * A;

  const strokeDashoffset = α * r;

  return (
    <Svg width={size} height={RFValue(230)}>
      <Defs>
        <LinearGradient id="grad" x1="0" y1="0" x2="100%" y2="0">
          <Stop offset="0" stopColor="#fbf0ce" />
          <Stop offset="1" stopColor="#dfaa6d" />
        </LinearGradient>
      </Defs>

      <Path
        stroke="#ef9837"
        strokeLinecap="round"
        fill="none"
        strokeDasharray={`${circumference}, ${circumference}`}
        {...{d, strokeWidth}}
      />

      <Path
        fill="none"
        strokeDasharray={`${circumference}, ${circumference}`}
        strokeDashoffset={strokeDashoffset}
        strokeLinecap="round"
        {...{d, strokeWidth}}
      />
      {content && (
        <Text
          fill="#fff"
          fontSize={32}
          fontFamily="Inter-Bold"
          textAnchor="middle"
          lengthAdjust="spacingAndGlyphs"
          x={size / 2}
          y={size / 2 + 10}>
          {numeral(content).format().slice(0, 10)}
        </Text>
      )}

      <Path
        stroke="url(#grad)"
        fill="none"
        opacity={0.8}
        strokeLinecap="round"
        strokeWidth={28}
        strokeDasharray={`${circumference}, ${circumference}`}
        {...{d: d2}}
      />
    </Svg>
  );
};
