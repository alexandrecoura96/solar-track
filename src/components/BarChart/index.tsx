import {
  Canvas,
  Path,
  Skia,
  Text,
  runTiming,
  useComputedValue,
  useFont,
  useValue,
} from '@shopify/react-native-skia';
import * as d3 from 'd3';
import numeral from 'numeral';
import React from 'react';
import {Button, Easing, StyleSheet, View} from 'react-native';
import {useSharedValue, withTiming} from 'react-native-reanimated';
import {BarChartDataProps, BarChartProps} from './types';

const GRAPH_MARGIN = 20;
const GRAPH_BAR_WIDTH = 8;

const CanvasHeight = 350;
const CanvasWidth = 350;
const graphHeight = CanvasHeight - 2 * GRAPH_MARGIN;
const graphWidth = CanvasWidth - 2;

export const BarChart = ({data}: BarChartProps) => {
  const font = useFont(require('../../assets/fonts/Inter-Regular.ttf'), 10);
  const animationState = useValue(0);
  const opacityAnimValue = useSharedValue(0);

  const xDomain = data.map((dataPoint: BarChartDataProps) => dataPoint.label);
  const xRange = [0, graphWidth];
  const x = d3.scalePoint().domain(xDomain).range(xRange).padding(1);

  const yDomain: number[] = [
    0,
    d3.max(
      data,
      (yBarChartDataProps: BarChartDataProps) => yBarChartDataProps.value,
    )!,
  ];

  const yRange = [0, graphHeight];
  const y = d3.scaleLinear().domain(yDomain).range(yRange);

  const animate = () => {
    animationState.current = 0;

    runTiming(animationState, 1, {
      duration: 1600,
      easing: Easing.inOut(Easing.exp),
    });

    opacityAnimValue.value = withTiming(1, {
      duration: 1600,
    });
  };

  const path = useComputedValue(() => {
    const newPath = Skia.Path.Make();

    data.forEach((dataPoint: BarChartDataProps) => {
      const rect = Skia.XYWHRect(
        x(dataPoint.label)! - GRAPH_BAR_WIDTH / 2,
        graphHeight,
        GRAPH_BAR_WIDTH,
        y(dataPoint.value * animationState.current) * -1,
      );

      const rrect = Skia.RRectXY(rect, 8, 8);
      newPath.addRRect(rrect);
    });

    return newPath;
  }, [animationState]);

  if (!font) {
    return <View />;
  }

  return (
    <View style={styles.container}>
      <Canvas style={styles.canvasTitle}>
        {data.map((dataPoint: BarChartDataProps) => (
          <Text
            key={dataPoint.label}
            color="#fff"
            font={font}
            x={x(dataPoint.label)! - 10}
            y={CanvasHeight - 60}
            text={numeral(dataPoint.value).format('0a')}
            opacity={opacityAnimValue}
          />
        ))}
      </Canvas>

      <Canvas style={styles.canvas}>
        <Path path={path} color="#e05f25" />

        {data.map((dataPoint: BarChartDataProps) => (
          <Text
            color="#fff"
            key={dataPoint.label}
            font={font}
            x={x(dataPoint.label)! - 10}
            y={CanvasHeight - 25}
            text={dataPoint.label}
          />
        ))}
      </Canvas>
      <Button title="Animate!" onPress={animate} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    flex: 0.5,
    backgroundColor: '#000',
  },
  canvasTitle: {
    height: CanvasHeight - 20,
    width: CanvasWidth,
    paddingBottom: 60,
  },
  canvas: {
    height: CanvasHeight,
    width: CanvasWidth,
  },
});
