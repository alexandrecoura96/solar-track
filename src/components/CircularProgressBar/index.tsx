import {
  Easing,
  SkFont,
  runTiming,
  useFont,
  useValue,
} from '@shopify/react-native-skia';
import React, {useCallback, useEffect} from 'react';
import {PixelRatio} from 'react-native';
import {DonutChart} from './DonutChart';
import {CircularProgressBarProps} from './types';

export function CircularProgressBar({percentage}: CircularProgressBarProps) {
  const font = useFont(require('../../assets/fonts/Inter-Black.ttf'), 16);
  const radius = PixelRatio.roundToNearestPixel(36);
  const STROKE_WIDTH = 5;
  const animationState = useValue(0);
  const targetPercentage = percentage / 100;

  const animateChart = useCallback(() => {
    animationState.current = 0;
    runTiming(animationState, targetPercentage, {
      duration: 1250,
      easing: Easing.inOut(Easing.cubic),
    });
  }, [animationState, targetPercentage]);

  useEffect(() => {
    animateChart();
  }, [animateChart]);

  if (!font) {
    return null;
  }

  return (
    <DonutChart
      radius={radius}
      strokeWidth={STROKE_WIDTH}
      percentageComplete={animationState}
      targetPercentage={targetPercentage}
      font={font as SkFont}
    />
  );
}
