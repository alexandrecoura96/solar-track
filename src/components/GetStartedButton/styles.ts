import {RFValue} from 'react-native-responsive-fontsize';
import {styled} from 'styled-components/native';

export const GetStarted = styled.TouchableOpacity`
  background-color: #1a1d1a;
  border-radius: ${RFValue(8)}px;
  flex-direction: row;
  justify-content: space-between;
  padding: ${RFValue(8)}px ${RFValue(16)}px;
  align-items: center;
`;

export const GetStartedText = styled.Text`
  color: #fff;
  font-size: ${RFValue(16)}px;
  font-family: 'InterTight-BlackItalic';
`;

export const GetStartedIconWrapper = styled.View`
  border-radius: ${RFValue(4)}px;
  padding: ${RFValue(8)}px;
  background-color: #aaf551;
  align-items: center;
  justify-content: center;
`;
