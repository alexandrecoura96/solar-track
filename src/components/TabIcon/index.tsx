import React from 'react';
import GraphicsIcon from '../../assets/svg/graphics.svg';
import HomeIcon from '../../assets/svg/home.svg';
import SettingsIcon from '../../assets/svg/settings.svg';

import {RFValue} from 'react-native-responsive-fontsize';
import {TabIconProps} from './types';

export const TabIcon = ({color}: TabIconProps) => {
  return {
    Home: <HomeIcon height={RFValue(24)} width={RFValue(24)} fill={color} />,
    Graphics: (
      <GraphicsIcon height={RFValue(24)} width={RFValue(24)} fill={color} />
    ),
    Settings: (
      <SettingsIcon height={RFValue(24)} width={RFValue(24)} fill={color} />
    ),
  };
};
