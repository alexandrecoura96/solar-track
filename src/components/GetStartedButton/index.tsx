import React from 'react';
import {RFValue} from 'react-native-responsive-fontsize';
import Play from '../../assets/svg/play.svg';
import {GetStarted, GetStartedIconWrapper, GetStartedText} from './styles';
import {GetStartedButtonProps} from './types';

export function GetStartedButton({title, ...rest}: GetStartedButtonProps) {
  return (
    <GetStarted {...rest} activeOpacity={0.8}>
      <GetStartedText>{title}</GetStartedText>
      <GetStartedIconWrapper>
        <Play height={RFValue(16)} width={RFValue(16)} />
      </GetStartedIconWrapper>
    </GetStarted>
  );
}
