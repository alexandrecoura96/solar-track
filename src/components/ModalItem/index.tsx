import React from 'react';
import {RFValue} from 'react-native-responsive-fontsize';
import Check from '../../assets/svg/check_circle.svg';
import {Container, Title} from './styles';
import {ModalItemProps} from './types';

export function ModalItem({
  title,
  selected,
  hasBorder = true,
  ...rest
}: ModalItemProps) {
  return (
    <Container hasBorder={hasBorder} activeOpacity={0.8} {...rest}>
      <Title>{title}</Title>
      {selected && (
        <Check fill="#1C1E1B" height={RFValue(20)} width={RFValue(20)} />
      )}
    </Container>
  );
}
