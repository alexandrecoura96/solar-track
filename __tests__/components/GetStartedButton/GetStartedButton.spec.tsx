import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {GetStartedButton} from '../../../src/components/GetStartedButton';

describe('GetStartedButton', () => {
  test('Should render tree of snapshot test in component', () => {
    const tree = renderer.create(<GetStartedButton title="title" />);

    expect(tree).toMatchSnapshot();
  });

  test('Fire event should be shot', () => {
    const handlePress = jest.fn();

    const tree = render(
      <GetStartedButton title="title" onPress={handlePress} />,
    );

    const button = tree.getByText('title');

    fireEvent.press(button);

    expect(handlePress).toBeCalled();
  });
});
