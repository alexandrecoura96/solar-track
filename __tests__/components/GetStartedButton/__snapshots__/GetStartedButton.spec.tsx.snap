// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetStartedButton Should render tree of snapshot test in component 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#1a1d1a",
      "borderBottomLeftRadius": 16,
      "borderBottomRightRadius": 16,
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
      "paddingBottom": 16,
      "paddingLeft": 31,
      "paddingRight": 31,
      "paddingTop": 16,
    }
  }
>
  <Text
    style={
      {
        "color": "#fff",
        "fontFamily": "InterTight-BlackItalic",
        "fontSize": 31,
      }
    }
  >
    title
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f4a460",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "justifyContent": "center",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <play.svg
      height={31}
      width={31}
    />
  </View>
</View>
`;
